src/lib/y2storage/proposal/lvm_helper.rb

```diff
@@ -0,0 +1,320 @@
+      # @see #useful_pv_space
+      #
+      # @param useful_size [DiskSize] size usable to allocate logical volumes
+      # @return [DiskSize] real size of the partition
+      def real_pv_size(useful_size)
+        useful_size.ceil(extent_size) + useless_pv_space
+      end
+
+      # Volumes groups that could be reused by the proposal, sorted by
+      # preference.
+      #
+      # @param devicegraph [Storage::Devicegraph]
+      # @return [Array<Storage::LvmVg>]
+      def reusable_volume_groups(devicegraph)
+        vgs = devicegraph.volume_groups
+        big_vgs, small_vgs = vgs.partition { |vg| total_size(vg) >= target_size }
+        # Use #vg_name to ensure stable sorting
+        big_vgs.sort_by! { |vg| [total_size(vg), vg.vg_name] }
+        small_vgs.sort_by! { |vg| [total_size(vg), vg.vg_name] }
+        small_vgs.reverse!
+        big_vgs + small_vgs
+      end
+
+      # Portion of a newly created physical volume that couldn't be used to
+      # allocate logical volumes because it would be reserved for LVM metadata.
+      def useless_pv_space
+        USELESS_PV_SPACE
+      end
+
+    protected
+
+      attr_reader :planned_volumes
+
+      def extent_size
+        if reused_volume_group
+          DiskSize.new(reused_volume_group.extent_size)
+        else
+          DEFAULT_EXTENT_SIZE
+        end
+      end
+
+      def target_size
+        planned_volumes.target_disk_size(rounding: extent_size)
+      end
+
+      def substract_reused_vg_size(size)
+        vg_size = total_size(reused_volume_group)
+        if vg_size < size
+          size - vg_size
+        else
+          DiskSize.zero
+        end
+      end
+
+      def find_reused_vg(devicegraph)
+        devicegraph.lvm_vgs.with(sid: reused_volume_group.sid).first
+      end
+
+      def create_volume_group(devicegraph)
+        name = available_name(DEFAULT_VG_NAME, devicegraph)
+        Storage::LvmVg.create(devicegraph, name)
+      end
+
+      # Extends the given volume group by adding as physical volumes the
+      # partitions in the given list.
+      #
+      # This method modifies the volume group received as first argument.
+      #
+      # @param volume_group [Storage::LvmVg] volume group to extend
+      # @param part_names [Array<String>] device names of the partitions
+      # @param devicegraph [Storage::Devicegraph] to fetch the partitions
+      def assign_physical_volumes!(volume_group, part_names, devicegraph)
+        devicegraph.partitions.with(name: part_names).each do |partition|
+          volume_group.add_lvm_pv(partition)
+        end
+      end
+
+      # Makes sure the given volume group has enough free extends to allocate
+      # all the planned volumes, by deleting the existing volume groups.
+      #
+      # This method modifies the volume group received as first argument.
+      #
+      # FIXME: the current implementation does not guarantee than the freed
+      # space is the minimum valid one.
```

Comment:
I'm wondering if this strategy is necessary at all or we'd be better of just deleting all LVs when reusing an existing VG. At least in the case when we don't fit into any free remaining space.

Imagine for example the VG was created by us and so probably has root+swap/boot+home. This strategy will free home (since root will be too small). I fear this might more or less not do what the user expects. At lease I'm having a hard time finding  a scenario where this strategy gives good results. :-)